File: golf/intel.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"math/big"
	"net"
	"os"
	"sort"
	"strings"
)

var F = strings.Fields("vpn proxy tor malware c2 scanner brute_force spammer compromised datacenter cdn anycast crawler bot cloud private_relay anonymizer mobile isp government")
var S = [...]float64{30, 25, 45, 95, 95, 55, 70, 65, 75, 15, 5, 0, 10, 40, 10, 15, 35, 0, 0, 0}
var L = [...]struct {
	t float64
	n string
}{{80, "critical"}, {60, "high"}, {35, "medium"}, {15, "low"}}

var le = binary.LittleEndian

type DB struct {
	V4s, V4e, V4m []uint32
	V4v           []uint16
	V6s, V6e, V6m []*big.Int
	V6v           []uint16
	Vt            [][4]uint32
	St            []string
	W             [20]float64
}

func Load(path string) *DB {
	d, _ := os.ReadFile(path)
	if le.Uint32(d) != 6 {
		panic("ver")
	}
	o := make([]int, 19)
	for i := range o {
		o[i] = int(le.Uint64(d[8+i*8:]))
	}
	cn, ln, v6n, valn, strn, off := o[0], o[1], o[2], o[3], o[4], o[5:]
	bi := make([]uint32, 65537)
	for i := range bi {
		bi[i] = le.Uint32(d[off[0]+i*4:])
	}
	N := cn + ln
	v4s, v4e, v4v := make([]uint32, N), make([]uint32, N), make([]uint16, N)
	for b, j := 0, uint32(0); b < 65536; b++ {
		for ; j < bi[b+1]; j++ {
			lo := uint32(le.Uint16(d[off[1]+int(j)*2:]))
			v4s[j] = uint32(b)<<16 | lo
			v4e[j] = v4s[j] + uint32(le.Uint16(d[off[2]+int(j)*2:]))
			v4v[j] = le.Uint16(d[off[3]+int(j)*2:])
		}
	}
	for i := 0; i < ln; i++ {
		v4s[cn+i] = le.Uint32(d[off[4]+i*4:])
		v4e[cn+i] = le.Uint32(d[off[5]+i*4:])
		v4v[cn+i] = le.Uint16(d[off[6]+i*2:])
	}
	idx := make([]int, N)
	for i := range idx {
		idx[i] = i
	}
	sort.SliceStable(idx, func(a, b int) bool { return v4s[idx[a]] < v4s[idx[b]] })
	ts, te, tv := make([]uint32, N), make([]uint32, N), make([]uint16, N)
	for i, k := range idx {
		ts[i], te[i], tv[i] = v4s[k], v4e[k], v4v[k]
	}
	v4s, v4e, v4v = ts, te, tv
	v4m := make([]uint32, N)
	var mx uint32
	for i, e := range v4e {
		if e > mx {
			mx = e
		}
		v4m[i] = mx
	}
	r6 := func(o int) []*big.Int {
		r := make([]*big.Int, v6n)
		for i := range r {
			b := make([]byte, 16)
			binary.BigEndian.PutUint64(b[:8], le.Uint64(d[o+i*16+8:]))
			binary.BigEndian.PutUint64(b[8:], le.Uint64(d[o+i*16:]))
			r[i] = new(big.Int).SetBytes(b)
		}
		return r
	}
	v6s, v6e := r6(off[7]), r6(off[8])
	v6v := make([]uint16, v6n)
	for i := range v6v {
		v6v[i] = le.Uint16(d[off[9]+i*2:])
	}
	v6m := make([]*big.Int, v6n)
	m := new(big.Int)
	for i, e := range v6e {
		if e.Cmp(m) > 0 {
			m = e
		}
		v6m[i] = m
	}
	vt := make([][4]uint32, valn)
	for i := range vt {
		for j := 0; j < 4; j++ {
			vt[i][j] = le.Uint32(d[off[10]+(i*4+j)*4:])
		}
	}
	sd := off[12]
	st := make([]string, strn)
	for i := range st {
		so, sl := int(le.Uint32(d[off[11]+i*8:])), int(le.Uint32(d[off[11]+i*8+4:]))
		st[i] = string(d[sd+so : sd+so+sl])
	}
	var w [20]float64
	if N > 0 {
		var c [20]int
		for _, vid := range v4v {
			b := vt[vid][0]
			for i := range F {
				if b&(1<<i) != 0 {
					c[i]++
				}
			}
		}
		for i := range F {
			cc := c[i]
			if cc == 0 {
				cc = 1
			}
			w[i] = S[i] * (1 + math.Log2(float64(N)/float64(cc))/24)
		}
	} else {
		w = S
	}
	return &DB{v4s, v4e, v4m, v4v, v6s, v6e, v6m, v6v, vt, st, w}
}

func r1(x float64) float64 { return math.Round(x*10) / 10 }

type M struct {
	Source   string   `json:"source"`
	Provider string   `json:"provider"`
	Range    string   `json:"range"`
	Flags    []string `json:"flags"`
	Weight   float64  `json:"weight"`
}

func ipFmt(b []byte) string { return net.IP(b).String() }

func Lookup(db *DB, ipStr string) map[string]any {
	addr := net.ParseIP(ipStr)
	matches := []M{}
	push := func(vid uint16, rng string) {
		b := db.Vt[vid][0]
		fl := []string{}
		mxw := 0.0
		for i, f := range F {
			if b&(1<<i) != 0 {
				fl = append(fl, f)
				if db.W[i] > mxw {
					mxw = db.W[i]
				}
			}
		}
		matches = append(matches, M{db.St[db.Vt[vid][2]], db.St[db.Vt[vid][1]], rng, fl, r1(mxw)})
	}
	if a4 := addr.To4(); a4 != nil {
		ip := binary.BigEndian.Uint32(a4)
		i := sort.Search(len(db.V4s), func(k int) bool { return db.V4s[k] > ip })
		for ; i > 0; i-- {
			j := i - 1
			if db.V4m[j] < ip {
				break
			}
			if db.V4e[j] >= ip {
				sb, eb := make([]byte, 4), make([]byte, 4)
				binary.BigEndian.PutUint32(sb, db.V4s[j])
				binary.BigEndian.PutUint32(eb, db.V4e[j])
				push(db.V4v[j], ipFmt(sb)+"-"+ipFmt(eb))
			}
		}
	} else {
		ip := new(big.Int).SetBytes(addr.To16())
		i := sort.Search(len(db.V6s), func(k int) bool { return db.V6s[k].Cmp(ip) > 0 })
		for ; i > 0; i-- {
			j := i - 1
			if db.V6m[j].Cmp(ip) < 0 {
				break
			}
			if db.V6e[j].Cmp(ip) >= 0 {
				sb, eb := make([]byte, 16), make([]byte, 16)
				db.V6s[j].FillBytes(sb)
				db.V6e[j].FillBytes(eb)
				push(db.V6v[j], ipFmt(sb)+"-"+ipFmt(eb))
			}
		}
	}
	sort.SliceStable(matches, func(i, j int) bool { return matches[i].Weight > matches[j].Weight })

	fi := map[string]int{}
	for i, f := range F {
		fi[f] = i
	}
	seen, src := map[string]bool{}, map[string]bool{}
	for _, m := range matches {
		for _, f := range m.Flags {
			seen[f] = true
		}
		src[m.Provider+"|"+m.Source] = true
	}
	ranked := make([]string, 0, len(seen))
	for f := range seen {
		ranked = append(ranked, f)
	}
	sort.SliceStable(ranked, func(i, j int) bool { return db.W[fi[ranked[i]]] > db.W[fi[ranked[j]]] })
	score := 0.0
	if len(ranked) > 0 {
		top, ex := db.W[fi[ranked[0]]], 0.0
		for _, f := range ranked[1:] {
			ex += db.W[fi[f]]
		}
		score = r1(math.Min(100, (top+ex*0.15)*(1+0.08*math.Log2(float64(len(src)+1)))))
	}
	verdict := "clean"
	if len(matches) > 0 {
		verdict = "minimal"
		for _, lv := range L {
			if score >= lv.t {
				verdict = lv.n
				break
			}
		}
	}
	allFlags, providers, sf, sp := []string{}, []string{}, map[string]bool{}, map[string]bool{}
	for _, m := range matches {
		for _, f := range m.Flags {
			if !sf[f] {
				sf[f] = true
				allFlags = append(allFlags, f)
			}
		}
		if m.Provider != "" && !sp[m.Provider] {
			sp[m.Provider] = true
			providers = append(providers, m.Provider)
		}
	}
	for i, p := range providers {
		if strings.EqualFold(p, "tor") {
			providers = append([]string{"Tor"}, append(append([]string{}, providers[:i]...), providers[i+1:]...)...)
			break
		}
	}
	reasons, top := ranked, ""
	if len(reasons) > 5 {
		reasons = reasons[:5]
	}
	if len(providers) > 0 {
		top = providers[0]
	}
	return map[string]any{
		"ip": ipStr, "found": len(matches) > 0, "verdict": verdict, "score": score,
		"detections": len(matches), "sources": len(src), "top_provider": top,
		"providers": providers, "flags": allFlags, "reasons": reasons, "matches": matches,
	}
}

func main() {
	ip := "8.8.8.8"
	if len(os.Args) > 1 {
		ip = os.Args[1]
	}
	b, _ := json.MarshalIndent(Lookup(Load("../intel.bin"), ip), "", "  ")
	fmt.Println(string(b))
}

File: golf/intel.nim
import std/[algorithm, json, math, net, os, strformat, strutils, tables]

const FLAGS = ["vpn","proxy","tor","malware","c2","scanner","brute_force",
  "spammer","compromised","datacenter","cdn","anycast","crawler","bot","cloud",
  "private_relay","anonymizer","mobile","isp","government"]
const SEV = [30.0,25.0,45.0,95.0,95.0,55.0,70.0,65.0,75.0,15.0,5.0,0.0,10.0,
  40.0,10.0,15.0,35.0,0.0,0.0,0.0]
const LEVELS = [(80.0,"critical"),(60.0,"high"),(35.0,"medium"),(15.0,"low")]

type U128 = tuple[hi, lo: uint64]

proc `<`(a, b: U128): bool =
  if a.hi != b.hi: a.hi < b.hi else: a.lo < b.lo
proc `<=`(a, b: U128): bool = not (b < a)
proc `>=`(a, b: U128): bool = not (a < b)

proc rdU16(d: string, o: int): uint16 = cast[ptr uint16](unsafeAddr d[o])[]
proc rdU32(d: string, o: int): uint32 = cast[ptr uint32](unsafeAddr d[o])[]
proc rdU64(d: string, o: int): uint64 = cast[ptr uint64](unsafeAddr d[o])[]

type DB = object
  v4s, v4e, v4m: seq[uint32]
  v4v: seq[uint16]
  v6s, v6e, v6m: seq[U128]
  v6v: seq[uint16]
  vt: seq[array[4, uint32]]
  st: seq[string]
  w: array[20, float]

proc round1(x: float): float = round(x * 10.0) / 10.0

proc fmtV4(x: uint32): string =
  let a = [uint8((x shr 24) and 0xff), uint8((x shr 16) and 0xff),
           uint8((x shr 8) and 0xff), uint8(x and 0xff)]
  $IpAddress(family: IPv4, address_v4: a)

proc fmtV6(x: U128): string =
  var a: array[16, uint8]
  for i in 0 ..< 8:
    a[i] = uint8((x.hi shr (56 - i*8)) and 0xff)
    a[8+i] = uint8((x.lo shr (56 - i*8)) and 0xff)
  $IpAddress(family: IPv6, address_v6: a)

proc ipToU32(a: IpAddress): uint32 =
  for b in a.address_v4: result = (result shl 8) or uint32(b)

proc ipToU128(a: IpAddress): U128 =
  for i in 0 ..< 8:
    result.hi = (result.hi shl 8) or uint64(a.address_v6[i])
    result.lo = (result.lo shl 8) or uint64(a.address_v6[8+i])

proc load(path: string): DB =
  let d = readFile(path)
  if rdU32(d, 0) != 6: quit("unsupported version")
  var o: array[19, int]
  for i in 0 ..< 19: o[i] = int(rdU64(d, 8 + i*8))
  let cn = o[0]; let ln = o[1]; let v6n = o[2]; let valn = o[3]; let strn = o[4]
  let off = o[5 .. 18]
  var bi = newSeq[uint32](65537)
  for i in 0 .. 65536: bi[i] = rdU32(d, off[0] + i*4)
  let n = cn + ln
  var s = newSeq[uint32](n)
  var e = newSeq[uint32](n)
  var v = newSeq[uint16](n)
  for b in 0 ..< 65536:
    for j in bi[b] ..< bi[b+1]:
      let ji = int(j)
      let lo = uint32(rdU16(d, off[1] + ji*2))
      s[ji] = (uint32(b) shl 16) or lo
      e[ji] = s[ji] + uint32(rdU16(d, off[2] + ji*2))
      v[ji] = rdU16(d, off[3] + ji*2)
  for i in 0 ..< ln:
    s[cn+i] = rdU32(d, off[4] + i*4)
    e[cn+i] = rdU32(d, off[5] + i*4)
    v[cn+i] = rdU16(d, off[6] + i*2)
  var idx = newSeq[int](n)
  for i in 0 ..< n: idx[i] = i
  idx.sort(proc(a, b: int): int = cmp(s[a], s[b]))
  result.v4s = newSeq[uint32](n)
  result.v4e = newSeq[uint32](n)
  result.v4v = newSeq[uint16](n)
  for i in 0 ..< n:
    result.v4s[i] = s[idx[i]]
    result.v4e[i] = e[idx[i]]
    result.v4v[i] = v[idx[i]]
  result.v4m = newSeq[uint32](n)
  var mx: uint32 = 0
  for i in 0 ..< n:
    if result.v4e[i] > mx: mx = result.v4e[i]
    result.v4m[i] = mx

  result.v6s = newSeq[U128](v6n)
  result.v6e = newSeq[U128](v6n)
  result.v6v = newSeq[uint16](v6n)
  for i in 0 ..< v6n:
    result.v6s[i] = (rdU64(d, off[7] + i*16 + 8), rdU64(d, off[7] + i*16))
    result.v6e[i] = (rdU64(d, off[8] + i*16 + 8), rdU64(d, off[8] + i*16))
    result.v6v[i] = rdU16(d, off[9] + i*2)
  result.v6m = newSeq[U128](v6n)
  var m6: U128 = (0'u64, 0'u64)
  for i in 0 ..< v6n:
    if m6 < result.v6e[i]: m6 = result.v6e[i]
    result.v6m[i] = m6

  result.vt = newSeq[array[4, uint32]](valn)
  for i in 0 ..< valn:
    for k in 0 ..< 4:
      result.vt[i][k] = rdU32(d, off[10] + i*16 + k*4)
  let sd = off[12]
  result.st = newSeq[string](strn)
  for i in 0 ..< strn:
    let so = int(rdU32(d, off[11] + i*8))
    let sl = int(rdU32(d, off[11] + i*8 + 4))
    result.st[i] = d[sd+so ..< sd+so+sl]

  if n > 0:
    var c: array[20, int]
    for vid in result.v4v:
      let b = result.vt[int(vid)][0]
      for i in 0 ..< 20:
        if (b and (1'u32 shl i)) != 0: inc c[i]
    for i in 0 ..< 20:
      result.w[i] = SEV[i] * (1.0 + log2(float(n) / float(max(c[i], 1))) / 24.0)
  else:
    for i in 0 ..< 20: result.w[i] = SEV[i]

proc upperU32(a: seq[uint32], ip: uint32): int =
  var lo = 0; var hi = a.len
  while lo < hi:
    let m = (lo + hi) div 2
    if a[m] > ip: hi = m else: lo = m + 1
  lo

proc upperU128(a: seq[U128], ip: U128): int =
  var lo = 0; var hi = a.len
  while lo < hi:
    let m = (lo + hi) div 2
    if ip < a[m]: hi = m else: lo = m + 1
  lo

type Match = object
  source, provider, rng: string
  flags: seq[string]
  weight: float

proc lookup(db: DB, ipStr: string): JsonNode =
  let addr0 = parseIpAddress(ipStr)
  let v4 = addr0.family == IPv4
  var matches: seq[Match]

  proc add(vid: uint16, rng: string) =
    let vt = db.vt[int(vid)]
    let b = vt[0]
    var fl: seq[string]
    var mxw = 0.0
    for i in 0 ..< 20:
      if (b and (1'u32 shl i)) != 0:
        fl.add(FLAGS[i])
        if db.w[i] > mxw: mxw = db.w[i]
    matches.add(Match(source: db.st[int(vt[2])], provider: db.st[int(vt[1])],
                      rng: rng, flags: fl, weight: round1(mxw)))

  if v4:
    let ip = ipToU32(addr0)
    var i = upperU32(db.v4s, ip)
    while i > 0:
      dec i
      if db.v4m[i] < ip: break
      if db.v4e[i] >= ip:
        add(db.v4v[i], &"{fmtV4(db.v4s[i])}-{fmtV4(db.v4e[i])}")
  else:
    let ip = ipToU128(addr0)
    var i = upperU128(db.v6s, ip)
    while i > 0:
      dec i
      if db.v6m[i] < ip: break
      if db.v6e[i] >= ip:
        add(db.v6v[i], &"{fmtV6(db.v6s[i])}-{fmtV6(db.v6e[i])}")

  matches.sort(proc(a, b: Match): int = cmp(b.weight, a.weight))

  var allFlags: seq[string]
  for m in matches:
    for f in m.flags:
      if f notin allFlags: allFlags.add(f)
  let fi = proc(f: string): int =
    for i, x in FLAGS: (if x == f: return i)
    -1
  var ranked = allFlags
  ranked.sort(proc(a, b: string): int = cmp(db.w[fi(b)], db.w[fi(a)]))

  var srcSet = initTable[string, bool]()
  for m in matches: srcSet[m.provider & "\x00" & m.source] = true
  let srcN = srcSet.len

  var score = 0.0
  if ranked.len > 0:
    let top = db.w[fi(ranked[0])]
    var ex = 0.0
    for i in 1 ..< ranked.len: ex += db.w[fi(ranked[i])]
    score = round1(min(100.0, (top + ex * 0.15) *
                       (1.0 + 0.08 * log2(float(srcN + 1)))))

  var verdict = "clean"
  if matches.len > 0:
    verdict = "minimal"
    for (t, n) in LEVELS:
      if score >= t: verdict = n; break

  var providers: seq[string]
  for m in matches:
    if m.provider.len > 0 and m.provider notin providers:
      providers.add(m.provider)
  var torIdx = -1
  for i, p in providers:
    if p.toLowerAscii == "tor": torIdx = i; break
  if torIdx >= 0:
    providers.delete(torIdx)
    providers.insert("Tor", 0)

  var reasons: seq[string]
  for i in 0 ..< min(5, ranked.len): reasons.add(ranked[i])
  let topProv = if providers.len > 0: providers[0] else: ""

  var ms = newJArray()
  for m in matches:
    ms.add(%* {"source": m.source, "provider": m.provider, "range": m.rng,
               "flags": m.flags, "weight": %m.weight})

  result = %* {
    "ip": ipStr, "found": matches.len > 0, "verdict": verdict,
    "score": %score, "detections": matches.len, "sources": srcN,
    "top_provider": topProv, "providers": providers, "flags": allFlags,
    "reasons": reasons, "matches": ms
  }

when isMainModule:
  let ip = if paramCount() > 0: paramStr(1) else: "8.8.8.8"
  let db = load("../intel.bin")
  echo pretty(lookup(db, ip), 2)
